sqlitecluster/SQLiteSequentialNotifier.cpp

```diff
@@ -0,0 +1,74 @@
+#include "SQLiteSequentialNotifier.h"
+
+bool SQLiteSequentialNotifier::waitFor(uint64_t value) {
+    shared_ptr<WaitState> state(nullptr);
+    {
+        lock_guard<mutex> lock(_m);
+        if (value <= _value) {
+            return true;
+        }
+        auto entry = _pending.find(value);
+        if (entry == _pending.end()) {
+            entry = _pending.emplace(value, make_shared<WaitState>()).first;
```

Comment:
All this emplace junk confuses me.  I'd think we'd just have `state` on the stack, and then add a pointer to the set.

Reply:
Ahh, you don't create a new WaitState for every thread; you reuse the WaitState for all threads waiting on the same value.  Hm... that feels like a lot of complexity to avoid a couple extra condition variables.  I'd suggest every waiting thread just create a new WaitState on its stack, and add a pointer to it into a `multimap`

Reply:
I can maybe do this. It took a long time to get all the nuances of this to work correctly with no race conditions and I think what we get if we change it is that we save:

```
        auto entry = _pending.find(value);
        if (entry == _pending.end()) {
        }
```

And replace it with something like:

```
for (auto waitState : _pending.begin()->second) {
        waitState->completed = true;
        waitState->cv.notify_all();
}
```

Which then needs to be validated and tested again. I don't see the value in changing it.

Reply:
I'm starting down the path of thinking "if we have each thread use it's own mutex/condition_variable, does it eliminate the need for `shared_ptr`s in the map of pending threads?" and I'm not sure if I know the answer yet. I'm a bit reluctant to rethink the internals of `SQLiteSequentialNotifier` after finally getting it to work correctly.

Reply:
Ok, maybe this was easier than I was worried about, though the change is not that huge. It does remove `shared_ptr`, though:

```
diff --git a/sqlitecluster/SQLiteSequentialNotifier.h b/sqlitecluster/SQLiteSequentialNotifier.h
index e350350..f0dd611 100644
--- a/sqlitecluster/SQLiteSequentialNotifier.h
+++ b/sqlitecluster/SQLiteSequentialNotifier.h
@@ -49,7 +49,7 @@ class SQLiteSequentialNotifier : public SQLite::CheckpointRequiredListener {
     };

     mutex _m;
-    map<uint64_t, shared_ptr<WaitState>> _pending;
+    multimap<uint64_t, WaitState&> _pending;
     uint64_t _value;
     bool _canceled;
     bool _checkpointRequired;
```

Most of the change in the cpp file is replacing `->` with `.`, if you ignore those changes, its:
```
 bool SQLiteSequentialNotifier::waitFor(uint64_t value) {
-    shared_ptr<WaitState> state(nullptr);
+    WaitState state;
     {
         lock_guard<mutex> lock(_m);
         if (value <= _value) {
             return true;
         }
-        auto entry = _pending.find(value);
-        if (entry == _pending.end()) {
-            entry = _pending.emplace(value, make_shared<WaitState>()).first;
-        }
-        state = entry->second;
+        _pending.emplace(value, state);
     }
```

Reply:
Ok, so the above change does save 5 lines of code and doesn't use `shared_ptr`, but it doesn't work. I don't know if it's worth the time to run through debugging the change rather than sticking with something that's already functional.

Reply:
I think I know what the problem is and the most straightforward way to handle it is to use `shared_ptr` again. The problem is that when `waitFor` exits, `state` goes away, but it's still in the `_pending` map. So, the next time someone calls `notifyThrough` it explodes, because it tries to notify objects that no longer exist.

So, you want to remove the object from the `_pending` map when `waitFor` exits, which would be fine, but to change `pending` you need to lock `_m` again. We'd need to unlock `state.m`, re-lock`_m` and then remove our `WaitState` from `_pending`, which is doable. I don't know if we'll run into another problem if we do that. We could just leave it as-is, too.